modules/signals/src/with-linked-state.ts

```diff
@@ -0,0 +1,101 @@
+import { linkedSignal, WritableSignal } from '@angular/core';
+
+import { toDeepSignal } from './deep-signal';
+import {
+  EmptyFeatureResult,
+  InnerSignalStore,
+  SignalsDictionary,
+  SignalStoreFeature,
+  SignalStoreFeatureResult,
+  StateSignals,
+} from './signal-store-models';
+import { isWritableSignal, STATE_SOURCE } from './state-source';
+import { Prettify } from './ts-helpers';
+
+type LinkedStateResult<
+  LinkedStateInput extends Record<
+    string | symbol,
+    WritableSignal<unknown> | (() => unknown)
+  >
+> = {
+  [K in keyof LinkedStateInput]: LinkedStateInput[K] extends WritableSignal<
+    infer V
+  >
+    ? V
+    : LinkedStateInput[K] extends () => infer V
+    ? V
+    : never;
+};
+
+/**
+ *
+ * @description
+ * Generates and adds the properties of a `linkedSignal`
+ * to the store's state.
+ *
+ * @usageNotes
+ * ```typescript
+ * const UserStore = signalStore(
+ *   withState({ options: [1, 2, 3] }),
+ *   withLinkedState(({ options }) => ({ selectOption: () => options()[0] ?? undefined }))
+ * );
+ * ```
+ *
+ * The resulting state is of type `{ options: number[], selectOption: number | undefined }`.
+ * Whenever the `options` signal changes, the `selectOption` will automatically update.
+ *
+ * For advanced use cases, `linkedSignal` can be called within `withLinkedState`:
+ *
+ * ```typescript
+ * const UserStore = signalStore(
+ *   withState({ id: 1 }),
+ *   withLinkedState(({ id }) => ({
+ *     user: linkedSignal({
+ *       source: id,
+ *       computation: () => ({ firstname: '', lastname: '' })
+ *     })
+ *   }))
+ * )
```

Comment:
we can add a more realistic example when explicit usage of `linkedSignal` is necessary:

```suggestion
 * type Option = { id: number; label: string };
 *
 * const UserStore = signalStore(
 *   withState({ options: new Array<Option>() }),
 *   withLinkedState(({ options }) => ({
 *     selectedOption: linkedSignal<Option[], Option>({
 *       source: options,
 *       computation: (newOptions, previous) => {
 *         const option = newOptions.find((o) => o.id=== previous?.value.id);
 *         return option ?? newOptions[0];
 *       },
 *     })
 *   }))
 * )
```